### PR TITLE
Feature/as data endpoint enhancements

### PR DESCRIPTION
**Pull request by elliot-sabitov**
feat: enhancing data endpoint annotation to support param, header and response schemas as well as tags and security to support a more extended usage:

```
@AsDataEndpoint(
    Namespace,
    HttpMethod,
    QuerySchema,
    BodySchema,
    ParamSchema,
    HeaderSchema,
    ResponseSchema,
    ["my-tag"],
    [AuthorizationSecurity]
  )
  handleEndpoint(
    request: FastifyRequest<{
      Querystring: QuerySchema;
      Body: BodySchema;
      Params: ParamSchema;
      Headers: HeaderSchema;
      Response: ResponseSchema;
    }>,
  ): Promise<ResponseSchema> {
    // ... 
}
```

As a test, created a temporary endpoint using the same schema as the 
```
 @AsDataEndpoint(
    Namespace.AuthorizationData,
    HttpMethod.Post,
    AuthorizationQuerySchema, // querySchema
    AuthorizationQuerySchema, // bodySchema
    AuthorizationQuerySchema, // paramSchema
    AuthorizationQuerySchema, // headerSchema
    AuthorizationQuerySchema, // responseSchema
    "my-tag",
    [AuthorizationSecurity],
    "Temporary test endpoint description"
  )
  getAuthorizationTest(
    request: FastifyRequest<{
      Querystring: AuthorizationQuerystring;
      Body: AuthorizationQuerystring;
      Params: AuthorizationQuerystring;
      Headers: AuthorizationQuerystring;
      Response: AuthorizationQuerystring;
    }>,
  ): Promise<AuthorizationData[]> {
    return this._module.authorizeRepository.readAllByQuerystring(request.query);
  }
```
and we are able to see the new endpoint in Swagger UI with the header, response, and param schemas defined:
Screenshot 2024-08-12 at 9 03 56 AM

and we can see the OpenAPI spec if we go to `/docs/json` which has the schemas defined:
```
"post": {
  "tags": [
    "my-tag"
  ],
  "description": "Temporary test endpoint description",
  "requestBody": {
    "content": {
      "application/json": {
        "schema": {
          "type": "object",
          "properties": {
            "idToken": {
              "type": "string"
            },
            "type": {
              "type": "string"
            }
          },
          "required": [
            "idToken",
            "type"
          ]
        }
      }
    },
    "required": true
  },
  "parameters": [
    {
      "schema": {
        "type": "string"
      },
      "in": "query",
      "name": "idToken",
      "required": true
    },
    {
      "schema": {
        "type": "string"
      },
      "in": "query",
      "name": "type",
      "required": true
    },
    {
      "schema": {
        "type": "string"
      },
      "in": "path",
      "name": "idToken",
      "required": true
    },
    {
      "schema": {
        "type": "string"
      },
      "in": "path",
      "name": "type",
      "required": true
    },
    {
      "schema": {
        "type": "string"
      },
      "in": "header",
      "name": "idToken",
      "required": true
    },
    {
      "schema": {
        "type": "string"
      },
      "in": "header",
      "name": "type",
      "required": true
    }
  ],
  "security": [
    {
      "authorization": []
    }
  ],
  "responses": {
    "200": {
      "description": "Default Response",
      "content": {
        "application/json": {
          "schema": {
            "type": "object",
            "properties": {
              "idToken": {
                "type": "string"
              },
              "type": {
                "type": "string"
              }
            },
            "required": [
              "idToken",
              "type"
            ]
          }
        }
      }
    }
  }
},
```

### CODE REVIEW

File: 02_Util/src/util/swagger.ts [hunk 1]
```diff
@@ -117,4 +117,119 @@ export function initSwagger(
...
+const registerFastifyAuth = async (server: FastifyInstance) => {
+  await server.register(FastifyAuth as any).after();
+  console.log((server as any).authorization);
+
+  server.decorate(
+    'authorization',
+    function (request: any, reply: any, done: any) {
+      try {
+        const token = getAuthorizationTokenFromRequest(request);
+        if (token === '123') {
```
Comment (ChrisWeissmann) [inline 02_Util/src/util/swagger.ts:184]: Hardcoded token. We shouldn't move that forward.

File: 02_Util/src/util/swagger.ts [hunk 2]
```diff
@@ -117,4 +117,119 @@ export function initSwagger(
...
+const registerFastifySwagger = (
+  systemConfig: SystemConfig,
+  server: FastifyInstance,
+) => {
+  server.register(fastifySwagger, {
+    openapi: {
+      info: {
+        title: 'CitrineOS Central System API',
+        description: 'Central System API for OCPP 2.0.1 messaging.',
+        version: '1.1.1',
```
Comment (ChrisWeissmann) [inline 02_Util/src/util/swagger.ts:207]: wondering if you can make this dynamic based on current version

File: 02_Util/src/util/swagger.ts [hunk 3]
```diff
@@ -117,4 +117,119 @@ export function initSwagger(
...
+      },
+      servers: [
+        {
+          url: `http://${systemConfig.centralSystem.host}:${systemConfig.centralSystem.port}`,
+          description: 'TODO',
```
Comment (ChrisWeissmann) [inline 02_Util/src/util/swagger.ts:212]: Can we fill out the todo?

**Review by ChrisWeissmann — CHANGES_REQUESTED** [review]

File: 02_Util/src/util/swagger.ts (hunk 2 above)
Comment (elliot-sabitov) [reply 02_Util/src/util/swagger.ts]: done 👍  and I also created this PR in OCPI to match - https://github.com/citrineos/citrineos-ocpi/pull/112

File: 02_Util/src/util/swagger.ts (hunk 3 above)
Comment (elliot-sabitov) [reply 02_Util/src/util/swagger.ts]: djusted

File: 02_Util/src/util/swagger.ts
Comment (thanaParis) [inline 02_Util/src/util/swagger.ts:1]: Are there any screenshots of the before/after for what these changes do to the swagger docs?

**Review by thanaParis — CHANGES_REQUESTED** [review]
@elliot-sabitov I'd like to try merging this in 1.4.0

File: 02_Util/src/util/swagger.ts
Comment (elliot-sabitov) [reply 02_Util/src/util/swagger.ts:1]: Adjusted the description to include screenshot showing example usage

File: 02_Util/src/util/swagger.ts [hunk 4]
```diff
@@ -117,4 +119,115 @@ export function initSwagger(
...
+  server.decorate(
+    'authorization',
+    function (request: any, reply: any, done: any) {
+      try {
+        const token = getAuthorizationTokenFromRequest(request);
+        console.log('Received authorization token', token);
+        done();
+      } catch (e) {
+        reply.code(HttpStatus.UNAUTHORIZED);
+        // .send(buildOcpiErrorResponse(HttpStatus.UNAUTHORIZED));
```
Comment (ChrisWeissmann) [inline 02_Util/src/util/swagger.ts:190]: nitpick: but if this isn't needed it should go :)

**Review by ChrisWeissmann — APPROVED** [review]

File: 02_Util/src/util/swagger.ts (hunk 4 above)
Comment (elliot-sabitov) [reply 02_Util/src/util/swagger.ts]: removed

**Pull request merged**